# Data Consistency Checks

## Contents

1- Performing Consistency Checks of the df_prods DataFrame 

2- Interpreting Statistical Information from the df_ords DataFrame

3- Checking for Mixed-Type Data in the df_ords DataFrame.

4- Running a Check for Missing Values in the df_ords dataframe and Reporting on it

5- Running a Check for Duplicate Values in the df_ords data and Reporting on it

6- Exporting the Cleaned df_prods and df_ords Data as “.csv” files in the “Prepared Data” Folder

In [1]:
import os
import pandas as pd
import numpy as np

In [2]:
path = r'/Users/calebhill/09:01:2023 - Instacart Basket Analysis'

In [11]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = 0)

## 1) Performing Consistency Checks of the df_prods DataFrame

In [14]:
df_prods.isnull().sum()

product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_prods.head()

product_name  aisle_id  \
product_id                                                                
1                                  Chocolate Sandwich Cookies        61   
2                                            All-Seasons Salt       104   
3                        Robust Golden Unsweetened Oolong Tea        94   
4           Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
5                                   Green Chile Anytime Sauce         5   

            department_id  prices  
product_id                         
1                      19     5.8  
2                      13     9.3  
3                       7     4.5  
4                       1    10.5  
5                      13     4.3

In [17]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
df_nan

product_name  aisle_id  department_id  prices
product_id                                              
34                  NaN       121             14    12.2
69                  NaN        26              7    11.8
116                 NaN        93              3    10.8
262                 NaN       110             13    12.1
525                 NaN       109             11     1.2
1511                NaN        84             16    14.3
1780                NaN       126             11    12.3
2240                NaN        52              1    14.2
2586                NaN       104             13    12.4
3159                NaN       126             11    13.1
3230                NaN       120             16    14.4
3736                NaN        41              8    14.8
4283                NaN        77              7    14.4
4790                NaN        91             16    14.5
38183               NaN        39             12    20.9
40440               NaN       120             16    14.8

In [20]:
df_prods.shape

(49693, 4)

In [21]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 4)

In [23]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_name  aisle_id  \
product_id                                                                
462                         Fiber 4g Gummy Dietary Supplement        70   
18458                                              Ranger IPA        27   
26808                    Black House Coffee Roasty Stout Beer        27   
35306       Gluten Free Organic Peanut Butter & Chocolate ...       121   
35491                                 Adore Forever Body Wash       127   

            department_id  prices  
product_id                         
462                    11     4.8  
18458                   5     9.2  
26808                   5    13.4  
35306                  14     6.8  
35491                  11     9.9

In [25]:
df_prods_clean.shape

(49677, 4)

In [26]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [28]:
df_prods_clean_no_dups.shape

(49672, 4)

In [29]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## 2) Interpreting Statistical Information from the df_ords DataFrame

In [30]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

## It looks like the order_hour_of_day column has a max value of 2.3 when it should be 24. 

## 3) Checking for Mixed-Type Data in the df_ords DataFrame.

In [51]:
#checks for mixed-type data by looping through each column with the provided parameters to check for inconsitencies
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(f'There are mixed values in column:{col}')
    else:
        print(f'No mixed values in column: {col}')

No mixed values in column: order_id
No mixed values in column: user_id
No mixed values in column: order_number
No mixed values in column: orders_day_of_week
No mixed values in column: order_hour_of_day
No mixed values in column: days_since_last_order


## 4) Running a Check for Missing Values in the df_ords dataframe and Reporting on it

In [43]:
df_ords.isnull().sum() #checks for missing values in the dataframe

order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [44]:
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull()== True] #puts NaN values and there corresponding columns in a subset

In [47]:
df_ords_nan #prints the subset 

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

## There are 206,209 missing values in the days_since_last_order column. It seems reasonable to assume that these are first time customers and if you look at the order number for each corresponding order_number column, the value is 1 

## Because this data is consistent with what you would expect for a first time customer I will leave the values as they are.

## 5) Running a Check for Duplicate Values in the df_ords data and Reporting on it


In [48]:
df_ords_dups = df_ords[df_ords.duplicated()] #puts any duplicates into a subset

In [49]:
df_ords_dups #looks at the subset to check if any duplicates were found

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

## No Duplicates Found

## 6) Exporting the Cleaned df_prods and df_ords Data as “.csv” files in the “Prepared Data” Folder 

In [1]:
#df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))
#(I already did this during the reading)

In [50]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))